internal/follow_handlers.go

```
@@ -0,0 +1,271 @@
+					imported++
+				}
+			}
+		}
+		if err := scanner.Err(); err != nil {
+			log.WithError(err).Error("error scanning feeds for import")
+			ctx.Error = true
+			ctx.Message = "Error importing feeds"
+			s.render("error", w, ctx)
+		}
+
+		if err := s.db.SetUser(ctx.Username, user); err != nil {
+			ctx.Error = true
+			ctx.Message = "Error importing feeds"
+			s.render("error", w, ctx)
+			return
+		}
+
+		ctx.Error = false
+		ctx.Message = fmt.Sprintf("Successfully imported %d feeds", imported)
+		s.render("error", w, ctx)
+		return
+	}
+}
+
+// UnfollowHandler ...
+func (s *Server) UnfollowHandler() httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+		ctx := NewContext(s.config, s.db, r)
+
+		nick := strings.TrimSpace(r.FormValue("nick"))
+
+		if nick == "" {
+			ctx.Error = true
+			ctx.Message = "No nick specified to unfollow"
+			s.render("error", w, ctx)
+			return
+		}
+
+		user := ctx.User
+		if user == nil {
+			log.Fatalf("user not found in context")
+		}
+
+		url, ok := user.Following[nick]
+		if !ok {
+			ctx.Error = true
+			ctx.Message = fmt.Sprintf("No feed found by the nick %s", nick)
+			s.render("error", w, ctx)
+			return
+		}
+
+		delete(user.Following, nick)
+
+		if err := s.db.SetUser(ctx.Username, user); err != nil {
+			ctx.Error = true
+			ctx.Message = fmt.Sprintf("Error unfollowing feed %s: %s", nick, url)
+			s.render("error", w, ctx)
+			return
+		}
+
+		if strings.HasPrefix(url, s.config.BaseURL) {
+			url = UserURL(url)
+			nick := NormalizeUsername(filepath.Base(url))
+			followee, err := s.db.GetUser(nick)
+			if err != nil {
+				log.WithError(err).Warnf("error loading user object for followee %s", nick)
+			} else {
+				if followee.Followers != nil {
+					delete(followee.Followers, user.Username)
+					if err := s.db.SetUser(followee.Username, followee); err != nil {
+						log.WithError(err).Warnf("error updating user object for followee %s", followee.Username)
```

should we return an error page on this error case?

---

Nah not here. I want to refactor all of this with #234 234 with an event-sourcing / pubsub pattern instead. All of this "special" `twtxt` bot handling code is a mess :/